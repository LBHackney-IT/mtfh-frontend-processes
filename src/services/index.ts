export { default as locale } from "./locale";
export { default as menu } from "./menu";
export * as processes from "./processes";
export * from "./config";
