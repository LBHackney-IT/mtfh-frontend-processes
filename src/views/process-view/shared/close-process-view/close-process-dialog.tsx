import React from "react";

import { Form, Formik } from "formik";
import * as Yup from "yup";

import { CloseProcessReasonForm } from "./close-process-reason-form";

import {
  Button,
  Center,
  Dialog,
  DialogActions,
  Spinner,
} from "@mtfh/common/lib/components";
import { useErrorCodes } from "@mtfh/common/lib/hooks";

export const closeProcessDialogSchema = (errorMessages: Record<string, string>) =>
  Yup.object({
    reasonForCancellation: Yup.string().required(errorMessages.W6),
  });

export type CloseProcessDialogFormData = Yup.Asserts<
  ReturnType<typeof closeProcessDialogSchema>
>;

export const CloseProcessDialog = ({
  isCloseProcessDialogOpen,
  setCloseProcessDialogOpen,
  setCloseProcessReason,
  mutate,
  isCancel,
}) => {
  const errorMessages = useErrorCodes();

  if (!errorMessages) {
    return (
      <Center>
        <Spinner />
      </Center>
    );
  }

  return (
    <Dialog
      isOpen={isCloseProcessDialogOpen}
      onDismiss={() => setCloseProcessDialogOpen(false)}
      title={`Are you sure you want to ${
        isCancel ? "cancel" : "close"
      } this process? You will have to begin the process from the start.`}
    >
      <Formik<CloseProcessDialogFormData>
        initialValues={{ reasonForCancellation: "" }}
        validateOnBlur={false}
        validateOnChange={false}
        validationSchema={closeProcessDialogSchema(errorMessages)}
        onSubmit={async (values) => {
          const { reasonForCancellation } = values;
          setCloseProcessReason(reasonForCancellation);
          mutate();
          setCloseProcessDialogOpen(false);
        }}
      >
        <Form noValidate id="cancel-process-form" className="cancel-process-form">
          <CloseProcessReasonForm isCancel={isCancel} />
          <DialogActions>
            <Button type="submit" data-testid="close-process-modal-submit">
              {isCancel ? "Cancel Process" : "Close case"}
            </Button>
            <Button variant="secondary" onClick={() => setCloseProcessDialogOpen(false)}>
              Back
            </Button>
          </DialogActions>
        </Form>
      </Formik>
    </Dialog>
  );
};
