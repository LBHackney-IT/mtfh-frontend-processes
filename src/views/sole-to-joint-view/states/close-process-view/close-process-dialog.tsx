import React from "react";

import { Form, Formik } from "formik";

import { CloseProcessReasonForm } from "./close-process-reason-form";

import { Button, Dialog, DialogActions } from "@mtfh/common/lib/components";

export const CloseProcessDialog = ({
  isCloseProcessDialogOpen,
  setCloseProcessDialogOpen,
  setCloseProcessReason,
  mutate,
  isCancel,
}) => {
  return (
    <Dialog
      isOpen={isCloseProcessDialogOpen}
      onDismiss={() => setCloseProcessDialogOpen(false)}
      title={`Are you sure you want to ${
        isCancel ? "cancel" : "close"
      } this process? You will have to begin the process from the start.`}
    >
      <Formik
        initialValues={{ reasonForCancellation: undefined }}
        onSubmit={async (values) => {
          const { reasonForCancellation } = values;
          setCloseProcessReason(reasonForCancellation);
          mutate();
          setCloseProcessDialogOpen(false);
        }}
      >
        <Form noValidate id="cancel-process-form" className="cancel-process-form">
          <CloseProcessReasonForm isCancel={isCancel} />
          <DialogActions>
            <Button type="submit" data-testid="close-process-modal-submit">
              {isCancel ? "Cancel Process" : "Close case"}
            </Button>
            <Button variant="secondary" onClick={() => setCloseProcessDialogOpen(false)}>
              Back
            </Button>
          </DialogActions>
        </Form>
      </Formik>
    </Dialog>
  );
};
