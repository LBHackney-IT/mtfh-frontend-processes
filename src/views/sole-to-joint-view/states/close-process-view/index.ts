export * from "./close-process-dialog";
export * from "./close-process-reason-form";
export * from "./close-process-view";
