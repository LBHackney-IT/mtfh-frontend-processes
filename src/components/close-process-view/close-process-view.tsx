import { Link as RouterLink } from "react-router-dom";

import { Form, Formik } from "formik";
import * as Yup from "yup";

import { locale } from "../../services";
import { Trigger } from "../../services/processes/types";
import { IProcess, ProcessComponentProps } from "../../types";

import { editProcess } from "@mtfh/common/lib/api/process/v1";
import {
  Button,
  Checkbox,
  Heading,
  InlineField,
  Link,
  List,
  Text,
} from "@mtfh/common/lib/components";

const closeProcessSchema = () =>
  Yup.object({
    hasNotifiedResident: Yup.boolean(),
  });

type CloseProcessFormData = Yup.Asserts<ReturnType<typeof closeProcessSchema>>;

interface CloseProcessViewProps extends ProcessComponentProps {
  processConfig: IProcess;
  setGlobalError?: any;
  closeProcessReason?: string;
  optional?: {
    trigger?: Trigger;
    nextStepsDescription?: boolean;
    closed?: boolean;
  };
}

export const CloseProcessView = ({
  process,
  processConfig,
  mutate,
  setGlobalError,
  closeProcessReason,
  optional = {
    nextStepsDescription: true,
  },
}: CloseProcessViewProps): JSX.Element => {
  const { closed, trigger, nextStepsDescription } = optional;
  const { state } = process.currentState;
  const { processClosed } = processConfig.states;

  return closed || state === processClosed.state ? (
    <>
      <Heading variant="h3">{locale.views.closeProcess.thankYouForConfirmation}</Heading>
      <List variant="bullets" style={{ marginLeft: "1em" }}>
        <Text size="sm">{locale.views.closeProcess.confirmationText}</Text>
      </List>
      <div style={{ marginTop: "1em" }}>
        <Link as={RouterLink} to="" variant="back-link">
          {locale.returnHomePage}
        </Link>
      </div>
    </>
  ) : (
    <>
      <Heading variant="h3">Next steps:</Heading>
      {nextStepsDescription && (
        <Text size="sm">
          The applicant is not eligible for sole to joint tenure. <br />
          This case will be closed once you have sent an outcome letter to the resident.
        </Text>
      )}
      <Formik<CloseProcessFormData>
        initialValues={{
          hasNotifiedResident: false,
        }}
        onSubmit={async () => {
          try {
            await editProcess({
              id: process.id,
              processTrigger: trigger || Trigger.CloseProcess,
              processName: process?.processName,
              etag: process.etag || "",
              formData: {
                hasNotifiedResident: true,
                Reason: closeProcessReason,
              },
              documents: [],
            });
            mutate();
          } catch (e: any) {
            if (setGlobalError) {
              setGlobalError(e.response?.status || 500);
            } else {
              console.log(e.response?.status || 500);
            }
          }
        }}
      >
        {({ values }) => (
          <Form noValidate id="close-process-form" className="mtfh-close-process-form">
            <InlineField name="hasNotifiedResident" type="checkbox">
              <Checkbox id="condition">
                {locale.views.closeProcess.outcomeLetterSent}
              </Checkbox>
            </InlineField>
            <Button type="submit" disabled={!values.hasNotifiedResident}>
              {locale.confirm}
            </Button>
          </Form>
        )}
      </Formik>
    </>
  );
};
