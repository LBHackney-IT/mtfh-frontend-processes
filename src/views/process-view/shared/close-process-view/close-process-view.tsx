import React from "react";
import { Link as RouterLink } from "react-router-dom";

import { Form, Formik } from "formik";
import * as Yup from "yup";

import { locale } from "../../../../services";
import { Trigger } from "../../../../services/processes/types";
import { IProcess, ProcessComponentProps } from "../../../../types";
import { isSameState } from "../../../../utils/processUtil";

import { editProcess } from "@mtfh/common/lib/api/process/v1";
import {
  Button,
  Checkbox,
  Heading,
  InlineField,
  Link,
  List,
  StatusBox,
  Text,
} from "@mtfh/common/lib/components";

const closeProcessSchema = () =>
  Yup.object({
    hasNotifiedResident: Yup.boolean(),
  });

type CloseProcessFormData = Yup.Asserts<ReturnType<typeof closeProcessSchema>>;

interface CloseProcessViewProps extends ProcessComponentProps {
  processConfig: IProcess;
  setGlobalError?: any;
  closeProcessReason?: string;
  statusBox?: boolean;
  trigger?: string;
}

export const CloseProcessView = ({
  process,
  processConfig,
  mutate,
  setGlobalError,
  closeProcessReason,
  statusBox = true,
  trigger,
}: CloseProcessViewProps): JSX.Element => {
  const { state } = process.currentState;
  const { processCancelled, processClosed, tenureUpdated } = processConfig.states;

  return (
    <>
      {statusBox && (
        <StatusBox
          variant="warning"
          title={
            isSameState(process.currentState, processClosed)
              ? locale.views.reviewDocuments.soleToJointClosed
              : locale.views.reviewDocuments.soleToJointWillBeClosed
          }
        >
          {closeProcessReason && (
            <Text style={{ marginTop: 15 }}>{closeProcessReason}</Text>
          )}
        </StatusBox>
      )}

      {[processClosed.state, processCancelled.state, tenureUpdated.state].includes(
        state,
      ) ? (
        <>
          <Heading variant="h3">
            {locale.views.closeProcess.thankYouForConfirmation}
          </Heading>
          <List variant="bullets" style={{ marginLeft: "1em" }}>
            <Text size="sm">{locale.views.closeProcess.confirmationText}</Text>
          </List>
          <div style={{ marginTop: "1em" }}>
            <Link as={RouterLink} to="" variant="back-link">
              {locale.returnHomePage}
            </Link>
          </div>
        </>
      ) : (
        <>
          <Heading variant="h3">Next steps:</Heading>
          <Formik<CloseProcessFormData>
            initialValues={{
              hasNotifiedResident: false,
            }}
            onSubmit={async () => {
              try {
                await editProcess({
                  id: process.id,
                  processTrigger: trigger || Trigger.CloseProcess,
                  processName: process?.processName,
                  etag: process.etag || "",
                  formData: {
                    hasNotifiedResident: true,
                    Reason: closeProcessReason,
                  },
                  documents: [],
                });
                mutate();
              } catch (e: any) {
                if (setGlobalError) {
                  setGlobalError(e.response?.status || 500);
                } else {
                  console.log(e.response?.status || 500);
                }
              }
            }}
          >
            {({ values }) => (
              <Form
                noValidate
                id="close-process-form"
                className="mtfh-close-process-form"
              >
                <InlineField name="hasNotifiedResident" type="checkbox">
                  <Checkbox id="condition">
                    {locale.views.closeProcess.outcomeLetterSent}
                  </Checkbox>
                </InlineField>
                <Button type="submit" disabled={!values.hasNotifiedResident}>
                  {locale.confirm}
                </Button>
              </Form>
            )}
          </Formik>
        </>
      )}
    </>
  );
};
