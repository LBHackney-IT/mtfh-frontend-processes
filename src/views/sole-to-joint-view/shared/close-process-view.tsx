import { useState } from "react";
import { Link as RouterLink } from "react-router-dom";

import { Form, Formik } from "formik";

import { locale } from "../../../services";
import { Trigger } from "../../../services/processes/types";
import { IProcess } from "../../../types";

import { Process, editProcess } from "@mtfh/common/lib/api/process/v1";
import { Button, Checkbox, Heading, Link, List, Text } from "@mtfh/common/lib/components";

interface CloseProcessViewProps {
  processConfig: IProcess;
  process: Process;
  mutate: () => void;
}

export const CloseProcessView = ({
  process,
  processConfig,
  mutate,
}: CloseProcessViewProps): JSX.Element => {
  const [confirmed, setConfirmed] = useState<boolean>(false);

  const { state } = process.currentState;
  const { processClosed } = processConfig.states;

  return state === processClosed.state ? (
    <>
      <Heading variant="h3">Outcome letter has been sent</Heading>
      <List variant="bullets" style={{ marginLeft: "1em" }}>
        <Text size="sm">
          This has been recorded and can be viewed in the case activity history
        </Text>
      </List>
      <div style={{ marginTop: "1em" }}>
        <Link as={RouterLink} to="" variant="back-link">
          Return to home page
        </Link>
      </div>
    </>
  ) : (
    <>
      <Heading variant="h3">Next steps:</Heading>
      <Formik
        initialValues={{}}
        onSubmit={async () => {
          try {
            await editProcess({
              id: process.id,
              processTrigger: Trigger.CloseProcess,
              processName: process?.processName,
              etag: process.etag || "",
              formData: {
                hasNotifiedResident: true,
              },
              documents: [],
            });
            mutate();
          } catch (e: any) {
            console.log(e.response?.status || 500);
          }
        }}
      >
        {() => (
          <Form noValidate id="breach-form" className="mtfh-breach-form">
            <Checkbox
              id="condition"
              checked={confirmed}
              onChange={() => setConfirmed(!confirmed)}
            >
              {locale.views.closeCase.outcomeLetterSent}
            </Checkbox>
            <Button type="submit" disabled={!confirmed}>
              {locale.confirm}
            </Button>
          </Form>
        )}
      </Formik>
    </>
  );
};
