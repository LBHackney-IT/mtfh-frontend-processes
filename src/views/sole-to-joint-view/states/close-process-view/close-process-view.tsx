import { useState } from "react";
import { Link as RouterLink } from "react-router-dom";

import { Form, Formik } from "formik";

import { locale } from "../../../../services";
import { Trigger } from "../../../../services/processes/types";
import { IProcess } from "../../../../types";

import { Process, editProcess } from "@mtfh/common/lib/api/process/v1";
import { Button, Checkbox, Heading, Link, List, Text } from "@mtfh/common/lib/components";

interface CloseProcessViewProps {
  processConfig: IProcess;
  process: Process;
  mutate: () => void;
  closeProcessReason?: string;
  optional?: {
    trigger?: Trigger;
    nextStepsDescription?: boolean;
  };
}

export const CloseProcessView = ({
  process,
  processConfig,
  mutate,
  closeProcessReason,
  optional = {
    nextStepsDescription: true,
  },
}: CloseProcessViewProps): JSX.Element => {
  const [confirmed, setConfirmed] = useState<boolean>(false);

  const { state } = process.currentState;
  const { processClosed } = processConfig.states;

  return state === processClosed.state ? (
    <>
      <Heading variant="h3">
        {locale.views.reviewDocuments.thankYouForConfirmation}
      </Heading>
      <List variant="bullets" style={{ marginLeft: "1em" }}>
        <Text size="sm">{locale.views.closeCase.confirmationText}</Text>
      </List>
      <div style={{ marginTop: "1em" }}>
        <Link as={RouterLink} to="" variant="back-link">
          {locale.returnHomePage}
        </Link>
      </div>
    </>
  ) : (
    <>
      <Heading variant="h3">Next steps:</Heading>
      {optional.nextStepsDescription && (
        <Text size="sm">
          The applicant is not eligible for sole to joint tenure. <br />
          This case will be closed once you have sent an outcome letter to the resident.
        </Text>
      )}
      <Formik
        initialValues={{}}
        onSubmit={async () => {
          try {
            await editProcess({
              id: process.id,
              processTrigger: optional.trigger || Trigger.CloseProcess,
              processName: process?.processName,
              etag: process.etag || "",
              formData: {
                hasNotifiedResident: true,
                Reason: closeProcessReason,
              },
              documents: [],
            });
            mutate();
          } catch (e: any) {
            console.log(e.response?.status || 500);
          }
        }}
      >
        {() => (
          <Form noValidate id="close-process-form" className="mtfh-close-process-form">
            <Checkbox
              id="condition"
              checked={confirmed}
              onChange={() => setConfirmed(!confirmed)}
            >
              {locale.views.closeProcess.outcomeLetterSent}
            </Checkbox>
            <Button type="submit" disabled={!confirmed}>
              {locale.confirm}
            </Button>
          </Form>
        )}
      </Formik>
    </>
  );
};
