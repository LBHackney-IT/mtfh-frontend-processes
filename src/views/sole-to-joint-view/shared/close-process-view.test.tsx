import React from "react";

import { render } from "@hackney/mtfh-test-utils";

import { processes } from "../../../services";
import { mockBreachChecksFailedState, mockProcessClosedState } from "../../../test-utils";
import { CloseProcessView } from "./close-process-view";

describe("close-process-view", () => {
  beforeEach(() => {
    jest.resetModules();
  });

  test("it renders CloseProcessView correctly for state=ProcessClosed", async () => {
    const { container } = render(
      <CloseProcessView
        processConfig={processes.soletojoint}
        process={mockProcessClosedState}
        mutate={() => {}}
      />,
      {
        url: "/processes/soletojoint/e63e68c7-84b0-3a48-b450-896e2c3d7735",
        path: "/processes/soletojoint/:processId",
      },
    );
    expect(container).toMatchSnapshot();
  });

  test("it renders CloseProcessView correctly for state!=ProcessClosed", async () => {
    const { container } = render(
      <CloseProcessView
        processConfig={processes.soletojoint}
        process={mockBreachChecksFailedState}
        mutate={() => {}}
      />,
      {
        url: "/processes/soletojoint/e63e68c7-84b0-3a48-b450-896e2c3d7735",
        path: "/processes/soletojoint/:processId",
      },
    );
    expect(container).toMatchSnapshot();
  });
});
