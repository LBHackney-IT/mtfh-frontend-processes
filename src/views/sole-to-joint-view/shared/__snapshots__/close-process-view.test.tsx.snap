// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`close-process-view it renders CloseProcessView correctly for state!=ProcessClosed 1`] = `
<div>
  <h3
    class="lbh-heading-h3"
  >
    Next steps:
  </h3>
  <form
    action="#"
    class="mtfh-breach-form"
    id="breach-form"
    novalidate=""
  >
    <div
      class="govuk-checkboxes__item"
    >
      <input
        class="govuk-checkboxes__input"
        id="condition"
        type="checkbox"
      />
      <label
        class="govuk-label govuk-checkboxes__label"
        for="condition"
      >
        I confirm that an outcome letter has been sent to the resident
      </label>
    </div>
    <button
      class="govuk-button lbh-button"
      disabled=""
      type="submit"
    >
      Confirm
    </button>
  </form>
</div>
`;

exports[`close-process-view it renders CloseProcessView correctly for state=ProcessClosed 1`] = `
<div>
  <h3
    class="lbh-heading-h3"
  >
    Outcome letter has been sent
  </h3>
  <ul
    class="lbh-list lbh-list--bullet"
    style="margin-left: 1em;"
  >
    <li>
      <p
        class="lbh-body-s"
      >
        This has been recorded and can be viewed in the case activity history
      </p>
    </li>
  </ul>
  <div
    style="margin-top: 1em;"
  >
    <a
      class="govuk-back-link lbh-back-link lbh-link--no-visited-state"
      href="/"
    >
      Return to home page
    </a>
  </div>
</div>
`;
